### DISCUSSION TITLE
Intersect two values in Go

### DISCUSSION THREAD

**Discussion by abaco** [cue-lang/cue#1507] [Q&A]
I'm experimenting with using CUE to handle complex data with user-defined schemas. I want to do things like store and retrieve values, link them to each other, etc. This involves some manipulation of CUE values, which is mostly easy stuff except for one operation that I don't know how to implement: the intersection of two values.

This is what it would do:
```go
v1 := ctx.CompileString(`a: [{x: 1, y: 2}, {x: 3, y: 4}], b: true, c: string`)
v2 := ctx.CompileString(`a: [...{x: int}], c: "foo", d: number`)
v3 := v1.Intersect(v2) // a: [{x: 1}, {x: 3}], c: "foo"
```

Can this be implemented in application code or is it the kind of feature that belongs in the library?
I've tried to read the implementation of `Value.Unify` but honestly I'm lost (I'm also new to Go).

**Comment by verdverm** [comment]
Recursively processing CUE values is something that should* be done in Go code. CUE intentionally prohibits recursion as part of being Turing _Incomplete_, or that you should not program (imperatively) in CUE.

The `pick` operator I wrote may also be intersection, or hopefully is a good reference for mutually recursing CUE values in Go. Note that the operators in Cuetils are things you can't or shouldn't do in pure CUE, though some of them (looking at you diff) might be implemented upstream in `cue` at some point. You may also want to keep an eye on the Query Proposal (#165) and if that could meet your needs.

https://github.com/hofstadter-io/cuetils/blob/main/structural/pick.go

*one can simulate unrolling recursion in pure CUE, but it is a huge performance hit.

Also, welcome to CUE!

**Reply by abaco** [reply]
Thank you, a `pick` operator is actually what I need right now!

However I've tried your code and it fails on [pick_test.go](https://github.com/hofstadter-io/cuetils/blob/main/structural/pick_test.go) with:
```go
  {pick: "a: [...int]", from: "a: [15, 54, 3], b: true", want: `
{
  a: [15, 54, 3]
}
  `
```
It only returns `a: []`.

I've dug a bit and `[...int]` has a `ListKind` but I don't know how to decompose it to get the `int` within it. `Value.List()` is empty and `Value.Expr()` returns the same value with no `op`. Isn't `...` a kind of operator? What sort of value is `[...int]`?

Anyway, I might be able to complete my prototype without ever recursing into this kind of lists. For now I'm content with knowing this is a well-posed problem that can be solved completely one day.

**Reply by verdverm** [reply]
List picking is not fully implemented

https://github.com/hofstadter-io/cuetils/blob/599ba5eb3c102dda8f4ee01da3a7e69c9fc113ed/structural/pick.go#L113
